### ISSUE TITLE
Download binaries for elm 0.19.1

### ISSUE THREAD

**Issue by fkowal** [tmohme/gradle-elm-plugin#6]
Elm started to publish binaries since version 0.19.1

https://github.com/elm/compiler/releases

It would be a greatly simplify CI (and overall) experience if there would exist a possibility to enable download these binaries. ie

```
elm {
    version = 0.19.1
    download = true
}
```
A similar feature exists in https://github.com/srs/gradle-node-plugin/blob/master/docs/node.md#configuring-the-plugin

**Comment by fkowal** [comment]
Here is something i tried (this is a similar approach the gradle-node-plugin uses.

The artifact is resolved properly.
But during download there is an issue:

https://github.com/elm/compiler/releases/download/0.19.1/binary-for-linux-64-bit.gz -> redirects to https://github-production-release-asset-2e65be.s3.amazonaws.com/....

Gradle follows the redirect but amazonaws doesn't seem to accept HEAD requests. Making a GET request to the redirect URL, works fine.

> Failed to get resource: HEAD. [HTTP HTTP/1.1 403 Forbidden: https://github-production-release-asset-2e65be.s3.amazonaws.com/
```
repositories {
    // artifact location: https://github.com/elm/compiler/releases/download/0.19.1/binary-for-linux-64-bit.gz
    ivy {
        url "https://github.com/elm/compiler/releases/download"
        patternLayout {
            artifact '[revision]/[module].[ext]'
        }
        metadataSources {
            ignoreGradleMetadataRedirection()
            artifact()
        }
        content {
            includeGroup "org.elm-lang"
        }
    }
}

configurations {
    elm
}

dependencies {
    elm "org.elm-lang:binary-for-linux-64-bit:0.19.1@gz"
}

```

**Comment by tmohme** [comment]
Thanks for your suggestion.
I'll have a look at it soon.

**Comment by tmohme** [comment]
I agree that it is a pretty useful feature, especially fo CI.
Just have to think about a sensible way for configuration & implementation.
Started to experiment on this . . . so hopefully we have an implementation in the near future.

**Comment by fkowal** [comment]
That is great new. Ill be glad to help if thats needed. So pls let me know

**Comment by tmohme** [comment]
Have very rough implementation as WIP on master. Not yet production grade.
@fkowal If you are still interested, pls have a look at the intended API (see `src/test/groovy/org/mohme/gradle/BuildLogicFunctionalTest.groovy`) and let me know your opinion.

**Comment by fkowal** [comment]
Looks very good. Great job!

**Comment by tmohme** [comment]
I'm quite busy lately and don't know when I will have enough time to finish this thoroughly.
Fortunately the main functionality is already stable so that I will release the current state as soon as the build succeeds in travis (currently there seems to be a problem installing `elm` in the build environment . . . being able to download `elm` dynamically is obviously a good idea ;) 
The remaining problems I see are the suboptimal internal structure (I can take care of this later) and  potentially another breaking API change when I restructure the optimize/debug flags so that the selection of an invalid combination will be impossible.

**Comment by tmohme** [comment]
Just released v4.0.0 with support for using a downloaded `elm` executable.

**Comment by fkowal** [comment]
That fantastic news.
By no means, don't feel any pressure. I'll get back with feedback /ux report ones I get to try it out. But it will also take me a while. Xmas and 4 birthdays.

Awesome work. Thanks so much for your oss contribution
